### PR TITLE
executor: optimize cursor read point get by reading through pessimistic lock cache

### PR DESCRIPTION
**Pull request by you06**
### What problem does this PR solve?

Issue Number: ref #36162

Problem Summary:

### What is changed and how it works?

For the point-get must lock the secondary key, lock it before reading, then later TiDB can read from pessimistic lock cache, this reduces one kv-get RPC. 

I test it by an edited sysbench update-non-index(wrap it in pessimistic txns) with the following options:

- clustered index off
- 32 threads
- 10,000,000 table size
- 1 table

#### Before

- TPS: 3174
- Latency(avg): 10.08ms
- Latency(.95): 16.71ms

Screen Shot 2022-07-12 at 10 11 01 PM

#### After

- TPS: 3509
- Latency(avg): 9.12ms
- Latency(.95): 15.27ms

Screen Shot 2022-07-12 at 11 14 33 PM

There are no kv-get requests.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
Optimize cursor read point get by reading through pessimistic lock cache.
```

### CODE REVIEW

**Comment by you06** [comment]
This optimization only works when the following conditions are hit:

- pessimistic transaction(do not include pessimistic auto commit)
- repeatable read
- non-clustered PK or unique index

Maybe we can push down the lock-if-exist mechanism(RC mode) into TiKV, then it's able to reduce one round RPC for RC, cc @TonsnakeLin.

File: executor/point_get.go [hunk 1]
```diff
@@ -246,10 +254,12 @@ func (e *PointGetExecutor) Next(ctx context.Context, req *chunk.Chunk) error {
 
 			// try lock the index key if isolation level is not read consistency
 			// also lock key if read consistency read a value
```
Comment (sticnarf) [inline executor/point_get.go:256]: Some part of the comments may be changed to adapt the new logic and some may be moved above L240.

File: executor/point_get.go [hunk 2]
```diff
@@ -237,6 +237,14 @@ func (e *PointGetExecutor) Next(ctx context.Context, req *chunk.Chunk) error {
 				return err
 			}
 
+			readFromCache := !e.ctx.GetSessionVars().IsPessimisticReadConsistency()
```
Comment (sticnarf) [inline executor/point_get.go:240]: I don't really get the name `readFromCache`... In the RC case, we cannot read from cache?

To my understanding, it is somewhat (whether to) `lockNonExistIdxKey`? (Correct me if I get it wrong)

**Review by sticnarf — APPROVED** [review]
LGTM

**Review by cfzjywxk — APPROVED** [review]

**Pull request merged**